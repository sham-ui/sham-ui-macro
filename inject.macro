/* eslint-env node */
'use strict';
const { createMacro } = require( 'babel-macros' );
const inject = require( './macros/inject' );

module.exports = createMacro( macro );

function macro( { references } ) {

    // After transpile with rollup in cjs import will:
    // var inject_macro = require( 'sham-ui/inject.macro' );
    // ...
    // inject_macro.inject();
    // ...
    // But for ES modules transpile to
    // import { inject } from 'sham-ui/inject.macro';
    ( references.default || references.inject ).forEach( referencePath => {
        inject( referencePath.findParent( x => x.isDecorator() ) );
    } );
}
