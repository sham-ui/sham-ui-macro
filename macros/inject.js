const t = require( 'babel-types' );

/**
 * Macro for `DI.resolve`
 * @function inject
 * @param {string} [name] Name of injected item
 *
 * @example
 * import inject from 'sham-ui-macro/inject.macro';
 * class Service {
 *     \@inject api;
 *     \@inject( 'sham-ui:store' ) uiStore;
 * }
 *
 *
 * // ↓ ↓ ↓ ↓ ↓ ↓
 *
 * class Service {
 *    get api() {
 *        return this.DI.resolve('api');
 *    }
 *
 *    get uiStore() {
 *        return this.DI.resolve('sham-ui:store');
 *    }
 *
 * }
 */

/**
 * @param {BabelPath} decoratorPath
 * @private
 */
module.exports = function inject( decoratorPath ) {
    const parentPath = decoratorPath.parentPath;
    if ( !parentPath.isClassProperty() ) {
        throw new Error( `Only ClassProperty allow, get "${parentPath.node.type}"` );
    }

    let name;
    if ( t.isCallExpression( decoratorPath.node.expression ) ) {
        name = decoratorPath.node.expression.arguments[ 0 ];
    } else {
        name = t.stringLiteral( parentPath.node.key.name );
    }

    parentPath.replaceWith(
        t.classMethod(
            'get',
            parentPath.node.key,
            [],
            t.blockStatement( [
                t.returnStatement(
                    t.callExpression(
                        t.memberExpression(
                            t.memberExpression(
                                t.thisExpression(),
                                t.identifier( 'DI' )
                            ),
                            t.identifier( 'resolve' )
                        ),
                        [ name ]
                    )
                )
            ] )
        )
    );
};
