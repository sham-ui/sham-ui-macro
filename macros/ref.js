const t = require( 'babel-types' );

/**
 * Macro for get/generate uniq reference
 * @function ref
 * @param {string} [name] Name of reference
 *
 * @example
 * // Config in package.json
 * // "babelMacros": {
 * //   "ref": {
 * //     "enabled": true,
 * //     "uniq": true
 * //   }
 * // }
 *
 * import { ref } from 'sham-ui-macro/ref.macro';
 *
 * const firstName = ref();
 * const lastName = ref();
 * const state = {
 *     [ firstName ]: 'John',
 *     [ lastName ]: 'Smith'
 * }
 * const fullName = state[ ref( 'firstName' ) ] + state[ ref( 'lastName' ) ];
 *
 * // ↓ ↓ ↓ ↓ ↓ ↓
 *
 * const firstName = 0;
 * const lastName = 1;
 * const state = {
 *     [ firstName ]: 'John',
 *     [ lastName ]: 'Smith'
 * }
 * const fullName = state[ 0 ] + state[ 1 ];
 */

/**
 * @param {BabelPath} callExpressionPath
 * @param {Object} config
 * @param {Object} state
 * @private
 */
module.exports = function ref( callExpressionPath, config, state ) {
    const { uniq = false } = config;

    let name;
    const args = callExpressionPath.node.arguments;
    if ( 0 === args.length  ) {
        const parentPath = callExpressionPath.parentPath;
        if ( !parentPath.isVariableDeclarator() ) {
            throw new Error( 'Pass name to ref' );
        }

        // Use variable as name
        name = parentPath.node.id.name;
    } else {
        name = args[ 0 ].value;
    }

    let identifier;
    if ( uniq ) {
        if ( !state.map.has( name ) ) {
            state.map.set(
                name,
                state.map.size
            );
        }
        identifier = t.NumericLiteral( state.map.get( name ) );
    } else {
        identifier = t.StringLiteral( name );
    }

    callExpressionPath.replaceWith(
        identifier
    );
};
