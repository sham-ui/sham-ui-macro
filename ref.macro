/* eslint-env node */
'use strict';
const { createMacro } = require( 'babel-macros' );
const ref = require( './macros/ref' );

const refState = {
    map: new Map()
};

module.exports = createMacro( macro, {
    configName: 'ref'
} );

function macro( { references, config } ) {
    const { enabled = false } = config;
    if ( enabled ) {

        // After transpile with rollup in cjs import will:
        // var ref_macro = require( 'sham-ui/ref.macro' );
        // ...
        // const a = ref_macro.ref();
        // ...
        // But for ES modules transpile to
        // import { ref } from 'sham-ui/ref.macro';
        ( references.default || references.ref ).forEach( referencePath => {
            ref(
                referencePath.findParent( x => x.isCallExpression() ),
                config,
                refState
            )
        } );
    } else {

        // Keep imports for process in app level
        return {
            keepImports: true
        }
    }
}
