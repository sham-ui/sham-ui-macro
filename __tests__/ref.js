/* eslint-env node */

const path = require( 'path' );
const pluginTester = require( 'babel-plugin-tester' );
const plugin = require( 'babel-plugin-macros' );

const projectRoot = path.join( __dirname, '../../' );

expect.addSnapshotSerializer( {
    print( val ) {
        return val.split( projectRoot ).join( '<PROJECT_ROOT>/' );
    },
    test( val ) {
        return typeof val === 'string';
    }
} );

pluginTester( {
    plugin,
    snapshot: true,
    babelOptions: {
        filename: __filename,
        parserOpts: {
            plugins: [
                'decorators',
                'classProperties'
            ]
        }
    },
    tests: [ {
        title: 'without args',
        code: `
            import { ref } from '../ref.macro';

            const foo = ref();
        `,
        snapshot: false
    }, {
        title: 'with args',
        code: `
            import { ref } from '../ref.macro';

            const foo = ref( 'bar' ) ;
        `
    }, {
        title: 'as key',
        code: `
            import { ref } from '../ref.macro';

            const state = { [ ref( 'foo' ) ]: true }; 
        `
    }, {
        title: 'with inject',
        code: `
            import { ref } from '../ref.macro';
            import { inject } from '../inject.macro'

            class FooComponent extends Template {
                @inject store;
            }

            const state = { [ ref( 'foo' ) ]: true };
        `
    }, {
        title: 'don\'t used',
        snapshot: false,
        code: `
            import '../ref.macro';
        `
    } ]
} );
