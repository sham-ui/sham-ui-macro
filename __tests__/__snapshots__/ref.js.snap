// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros as key: as key 1`] = `

import { ref } from '../ref.macro';

const state = { [ ref( 'foo' ) ]: true };

      ↓ ↓ ↓ ↓ ↓ ↓

import { ref } from '../ref.macro';

const state = { [ref('foo')]: true };

`;

exports[`macros with args: with args 1`] = `

import { ref } from '../ref.macro';

const foo = ref( 'bar' ) ;

      ↓ ↓ ↓ ↓ ↓ ↓

import { ref } from '../ref.macro';

const foo = ref('bar');

`;

exports[`macros with inject: with inject 1`] = `

import { ref } from '../ref.macro';
import { inject } from '../inject.macro'

class FooComponent extends Template {
    @inject store;
}

const state = { [ ref( 'foo' ) ]: true };

      ↓ ↓ ↓ ↓ ↓ ↓

import { ref } from '../ref.macro';


class FooComponent extends Template {
    get store() {
        return this.DI.resolve('store');
    }

}

const state = { [ref('foo')]: true };

`;
